src/rspq/rspq.c

```diff
@@ -0,0 +1,1235 @@
+                 SP_WSTATUS_CLEAR_SIG_HIGHPRI_REQUESTED |
+                 SP_WSTATUS_CLEAR_SIG_MORE;
+
+    MEMORY_BARRIER();
+
+    // Off we go!
+    rsp_run_async();
+
+    // Disable INTR_ON_BREAK as that it is not useful in the RSPQ engine, and
+    // might even cause excessive interrupts.
+    // It was turned on by rsp_run_async.
+    *SP_STATUS = SP_WSTATUS_CLEAR_INTR_BREAK;
+}
+
+/** @brief Initialize a rspq_ctx_t structure */
+static void rspq_init_context(rspq_ctx_t *ctx, int buf_size)
+{
+    memset(ctx, 0, sizeof(rspq_ctx_t));
+    ctx->buffers[0] = malloc_uncached(buf_size * sizeof(uint32_t));
+    ctx->buffers[1] = malloc_uncached(buf_size * sizeof(uint32_t));
+    memset(ctx->buffers[0], 0, buf_size * sizeof(uint32_t));
+    memset(ctx->buffers[1], 0, buf_size * sizeof(uint32_t));
+    ctx->buf_idx = 0;
+    ctx->buf_size = buf_size;
+    ctx->cur = ctx->buffers[0];
+    ctx->sentinel = ctx->cur + buf_size - RSPQ_MAX_COMMAND_SIZE;
+}
+
+static void rspq_close_context(rspq_ctx_t *ctx)
+{
+    free_uncached(ctx->buffers[1]);
+    free_uncached(ctx->buffers[0]);
+}
+
+void rspq_init(void)
+{
+    // Do nothing if rspq_init has already been called
+    if (rspq_initialized)
+        return;
+
+    rspq_ctx = NULL;
+    rspq_cur_pointer = NULL;
+    rspq_cur_sentinel = NULL;
+
+    // Allocate RSPQ contexts
+    rspq_init_context(&lowpri, RSPQ_DRAM_LOWPRI_BUFFER_SIZE);
+    lowpri.sp_status_bufdone = SP_STATUS_SIG_BUFDONE_LOW;
+    lowpri.sp_wstatus_set_bufdone = SP_WSTATUS_SET_SIG_BUFDONE_LOW;
+    lowpri.sp_wstatus_clear_bufdone = SP_WSTATUS_CLEAR_SIG_BUFDONE_LOW;
+
+    rspq_init_context(&highpri, RSPQ_DRAM_HIGHPRI_BUFFER_SIZE);
+    highpri.sp_status_bufdone = SP_STATUS_SIG_BUFDONE_HIGH;
+    highpri.sp_wstatus_set_bufdone = SP_WSTATUS_SET_SIG_BUFDONE_HIGH;
+    highpri.sp_wstatus_clear_bufdone = SP_WSTATUS_CLEAR_SIG_BUFDONE_HIGH;
+
+    // Start in low-priority mode
+    rspq_switch_context(&lowpri);
+
+    // Load initial settings
+    memset(&rspq_data, 0, sizeof(rsp_queue_t));
+    rspq_data.rspq_dram_lowpri_addr = PhysicalAddr(lowpri.cur);
+    rspq_data.rspq_dram_highpri_addr = PhysicalAddr(highpri.cur);
+    rspq_data.rspq_dram_addr = rspq_data.rspq_dram_lowpri_addr;
+    rspq_data.tables.overlay_descriptors[0].state = PhysicalAddr(&dummy_overlay_state);
+    rspq_data.tables.overlay_descriptors[0].data_size = sizeof(uint64_t);
+    rspq_data.current_ovl = 0;
+    
+    // Init syncpoints
+    rspq_syncpoints_genid = 0;
+    rspq_syncpoints_done = 0;
+
+    // Init blocks
+    rspq_block = NULL;
+    rspq_is_running = false;
+
+    // Activate SP interrupt (used for syncpoints)
+    register_SP_handler(rspq_sp_interrupt);
+    set_SP_interrupt(1);
+
+    rspq_initialized = 1;
+
+    rspq_start();
+}
+
+/** @brief Stop the RSP queue engine in the RSP */
+static void rspq_stop(void)
+{
+    MEMORY_BARRIER();
+    *SP_STATUS = SP_WSTATUS_SET_HALT;
+    MEMORY_BARRIER();
+
+    rspq_is_running = 0;
```

Comment:
How many cycles does it take for the RSP to actually halt, I think it is not possible to know e.g if a command was already running? Might be a good idea to warn about doing a `rspq_stop` + `rspq_start` in close succession might not work correctly. Would adding a a syncpoint help here?

Reply:
`rspq_stop` is only used to tear down the entire system when calling `rspq_close`, so it should be on the user to make sure that all commands they care about have finished executing. Also, `rspq_close` will probably never be called in an actual game because there isn't really any reason to do so. The system can be used continuously between scenes, and synchronisation can be done with syncpoints by the user. We only added this function for completeness's sake and to make the tests work basically.

Calling `rspq_init` straight after `rspq_close` shouldn't really cause issues because the former will always perform the entire setup necessary to bootstrap the rspq system.

And just in case you might have missed it: `rspq_start` and `rspq_stop` are not public API :)